File: api/controllers/AuthController.php
<?php
/**
 * Auth Controller - Login, Register, Forgot Password
 */
require_once __DIR__ . '/../config/database.php';
require_once __DIR__ . '/../models/User.php';
require_once __DIR__ . '/../utils/jwt_helper.php';
require_once __DIR__ . '/../core/Response.php';

class AuthController {
    private $db;
    private $user;

    public function __construct() {
        $database = new Database();
        $this->db = $database->getConnection();
        $this->user = new User($this->db);
    }

    public function login() {
        $data = json_decode(file_get_contents("php://input"));
        if (empty($data->email) || empty($data->password)) {
            Response::error("Dados incompletos.", 400);
            return;
        }

        $email = htmlspecialchars(strip_tags($data->email));
        $password = $data->password;

        $query = "SELECT id, name, password, email, avatar_url, weight_lost FROM users WHERE email = ? LIMIT 0,1";
        $stmt = $this->db->prepare($query);
        $stmt->bindParam(1, $email);
        $stmt->execute();

        if ($stmt->rowCount() > 0) {
            $row = $stmt->fetch(PDO::FETCH_ASSOC);
            if (password_verify($password, $row['password'])) {
                $token = JWT_Helper::create(['user_id' => (int) $row['id']]);
                Response::json([
                    "status" => "success",
                    "message" => "Login realizado com sucesso.",
                    "token" => $token,
                    "user" => [
                        "id" => $row['id'],
                        "name" => $row['name'],
                        "email" => $row['email'],
                        "avatar_url" => $row['avatar_url'],
                        "weight_lost" => $row['weight_lost']
                    ]
                ], 200);
            } else {
                Response::error("Senha incorreta!", 401);
            }
        } else {
            Response::error("Email nao encontrado.", 404);
        }
    }

    public function register() {
        $data = json_decode(file_get_contents("php://input"));
        if (empty($data->name) || empty($data->email) || empty($data->password)) {
            Response::error("Preencha todos os campos.", 400);
            return;
        }

        if (!filter_var($data->email, FILTER_VALIDATE_EMAIL)) {
            Response::error("Formato de email invalido.", 400);
            return;
        }

        $this->user->email = $data->email;
        if ($this->user->emailExists()) {
            Response::error("Este email ja esta registado.", 400);
            return;
        }

        $this->user->name = $data->name;
        $this->user->password = password_hash($data->password, PASSWORD_BCRYPT);

        if ($this->user->create()) {
            Response::json([
                "status" => "success",
                "message" => "Conta criada com sucesso."
            ], 201);
        } else {
            Response::error("Nao foi possivel criar a conta.", 503);
        }
    }

    public function forgotPassword() {
        $data = json_decode(file_get_contents("php://input"));
        if (empty($data->email)) {
            Response::error("Por favor insira o email.", 400);
            return;
        }

        $this->user->email = $data->email;
        if ($this->user->emailExists()) {
            Response::json([
                "status" => "success",
                "message" => "Email de recuperacao enviado!"
            ], 200);
        } else {
            Response::error("Este email nao esta registado na nossa base de dados.", 404);
        }
    }
}

File: app/lib/controllers/auth_controller.dart
import 'package:dio/dio.dart';
import '../core/api_client.dart';

/// Auth Controller - Login, Register, Forgot Password logic
class AuthController {
  final ApiClient _api = ApiClient();

  Future<bool> login(String email, String password) async {
    final response = await _api.dio.post(
      'auth_login.php',
      data: {'email': email.trim(), 'password': password},
    );

    if (response.statusCode == 200) {
      final data = response.data;
      if (data['status'] == 'success') {
        if (data['token'] != null) ApiClient.authToken = data['token'];
        if (data['user'] != null) {
          ApiClient.userName = data['user']['name'];
          ApiClient.userAvatarUrl = data['user']['avatar_url'];
        }
        return true;
      }
    }
    return false;
  }

  Future<String?> loginWithError(String email, String password) async {
    try {
      final ok = await login(email, password);
      return ok ? null : 'Erro de login';
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) return 'Email ou palavra-passe incorretos.';
      if (e.response?.statusCode == 404) return 'Conta não encontrada.';
      return e.response?.data['message']?.toString() ?? 'Erro no servidor.';
    } catch (_) {
      return 'Ocorreu um erro inesperado.';
    }
  }

  Future<String?> register(String name, String email, String password) async {
    try {
      final response = await _api.dio.post('auth_register.php', data: {
        'name': name,
        'email': email.trim(),
        'password': password,
      });

      if (response.statusCode == 201) {
        await login(email, password);
        return null;
      }
      return response.data['message']?.toString() ?? 'Erro ao registar';
    } on DioException catch (e) {
      return e.response?.data['message']?.toString() ?? 'Erro ao registar';
    } catch (e) {
      return 'Erro ao registar: $e';
    }
  }

  Future<String?> forgotPassword(String email) async {
    try {
      final response = await _api.dio.post(
        'auth_forgot_password.php',
        data: {'email': email.trim()},
      );

      if (response.statusCode == 200) return null;
      return response.data['message']?.toString() ?? 'Erro ao enviar email.';
    } on DioException catch (e) {
      if (e.response?.statusCode == 404) {
        return 'Este email não está registado na nossa aplicação.';
      }
      return e.response?.data['message']?.toString() ?? 'Erro no servidor.';
    } catch (_) {
      return 'Erro ao enviar email.';
    }
  }
}
